File: 2501-longest-square-streak-in-an-array/2501-longest-square-streak-in-an-array.py
class Solution:
    def longestSquareStreak(self, nums: List[int]) -> int:
        num_map = dict()
        result = -1
        nums.sort()

        for num in nums:
            sqrt = isqrt(num)

            if sqrt * sqrt == num and sqrt in num_map:
                num_map[num] = num_map[sqrt] + 1
                result = max(result, num_map[num])
            else:
                num_map[num] = 1

        return result

File: 2501-longest-square-streak-in-an-array/2501-longest-square-streak-in-an-array.java
class Solution {
    public int longestSquareStreak(int[] nums) {
        Map<Integer, Integer> numMap = new HashMap<>();
        Arrays.sort(nums);
        int result = -1;

        for (int num: nums) {
            int sqrt = (int) Math.sqrt(num);

            if (sqrt * sqrt == num && numMap.containsKey(sqrt)) {
                numMap.put(num, numMap.get(sqrt) + 1);
                result = Math.max(result, numMap.get(num));
            } else {
                numMap.put(num, 1);
            }
        }

        return result;
    }
}